File: benchmark/latency.py
#!/usr/env python3

import time

import rclpy
from sensor_msgs.msg import Image
from builtin_interfaces.msg import Time

rclpy.init()
node = rclpy.create_node("canlab_benchmark");

def from_ros_time(ros_time: Time) -> float:
  return ros_time.sec + ros_time.nanosec / 1000000000

def on_image(cam_id: int, image: Image) -> float:
  local_time = from_ros_time(node.get_clock().now().to_msg());
  cam_time = from_ros_time(image.header.stamp)
  print(f"cam_{cam_id} latency = {(local_time - cam_time) * 1000000:.3f}us");

for i in range(4):
  node.create_subscription(Image, f"cam_{i}/image_raw", lambda image, i=i: on_image(i, image), 10)

rclpy.spin(node)

File: benchmarks/latency.cpp
/*
 * Copyright (C) 2022 Can-lab Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <ros/ros.h>

#include <iomanip>
#include <sensor_msgs/Image.h>

//==============================================================================
int main(int argc, char* argv[])
{
  ros::init(argc, argv, "clpe");

  const std::size_t cam_count = 4;

  auto node = ros::NodeHandle("canlab_benchmark_latency");
  std::array<std::size_t, cam_count> cam_sub_count;
  // Latency in millis
  std::array<double, cam_count> avg_latencies;
  for (auto i = 0; i < cam_count; ++i)
  {
    cam_sub_count[i] = 1;
    avg_latencies[i] = 0.0;
  }

  std::array<ros::Subscriber, cam_count> subs;
  using sensor_msgs::Image;
  for (int i = 0; i < cam_count; ++i)
  {
    subs[i] = node.subscribe<Image>("/clpe_node/cam_" + std::to_string(i) + "/image_raw", 10,
                                    [i, &node, &cam_sub_count, &avg_latencies](const Image::ConstPtr& image) {
                                      const auto local_time = ros::Time::now().toNSec();
                                      const auto cam_time = image->header.stamp.toNSec();
                                      const auto diff_ns = local_time - cam_time;
                                      const double latency = diff_ns / 1000000.0;
                                      avg_latencies[i] =
                                          (avg_latencies[i] * cam_sub_count[i] + latency) / (cam_sub_count[i] + 1);
                                      cam_sub_count[i]++;
                                      std::cout << "cam_" << i << " latency = " << std::fixed << std::setprecision(3)
                                                << avg_latencies[i] << "ms" << std::endl;
                                    });
  }

  ros::spin();

  return 0;
}
